src/components/RemoteImage.js

```diff
@@ -0,0 +1,39 @@
+import React from 'react'
+import memoize from 'memoize-one'
+import { noop } from '../utils'
+
+// Tries to load an image, while displaying an alternative.
+// Use `children` or `render` to render its children.
+class RemoteImage extends React.Component {
+  static defaultProps = {
+    src: '',
+
+    render: null, // render is an alias of children and takes priority
+    children: noop,
+  }
+  state = { exists: false }
+
+  updateSrc = memoize(src => {
+    this.setState(() => {
+      this._img = new Image()
+      this._img.onload = () => {
+        this.setState({ exists: true })
+      }
+      this._img.src = src
+
+      return { exists: false }
+    })
+  })
+
+  render() {
+    const { exists } = this.state
+    const { src, render, children } = this.props
+
+    // memoized, safe to be called here
+    this.updateSrc(src)
```

Comment:
Yeah I was trying to avoid lifecycle methods, since React is moving [towards that direction](https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html).

But Suspense is not there yet and I don’t mind either way, so let’s move it to`componentDidUpdate()`.

Reply:
This bit seems pretty tricky to handle, since it's async. We'll likely need to do something on `componentWillUnmount()` to make sure it doesn't leak.

Reply:
I’m pretty sure React removes any reference added to the component when unmounted (otherwise we would have to do it every time we use `ref`)… let me check again 🤔

Reply:
Does React remove the listeners too? That's pretty impressive, but otherwise, it still sounds like a race condition.

Reply:
Nope, you’re right 😱 👍